rest/admin_api.go

```diff
@@ -915,19 +915,28 @@ func (h *handler) handlePutDbConfigImportFilter() error {
 // In non-persistent mode, the endpoint just removes the database from the node.
 func (h *handler) handleDeleteDB() error {
 	h.assertAdminOnly()
+	dbName := h.PathVar("olddb")
+
+	var bucket string
+
+	dbContext, _ := h.server.GetDatabase(h.ctx(), dbName)
+
+	if dbContext != nil {
+		bucket = dbContext.Bucket.GetName()
+	} else if h.server.persistentConfig {
+		bucket, _ = h.server.bucketNameFromDbName(dbName)
+	}
```

Comment:
I'm not sure why we have to get bucket name for non-persistent config codepath (and do the `bucket == ""` check on line 938)? Is it nessesary?

Reply:
And to be specific, I was thinking of something like this:

- `handleDeleteDB` treats the persistent config path as a special case inside the existing if block, rather than introducing extra variables (like `bucket`/`dbContext`) that aren't set for some code-paths.
- `bucketNameFromDbName` can avoid a bunch of work looking up db config in buckets by first looking to see if we have a matching database loaded
  - this also has the side-effect/benefit that we don't need the GetDatabase call inside `handleDeleteDB`, because the only reason we needed that was to grab bucket name anyway.

```diff
diff --git a/rest/admin_api.go b/rest/admin_api.go
index ea00a21a..9d6f0411 100644
--- a/rest/admin_api.go
+++ b/rest/admin_api.go
@@ -962,27 +962,28 @@ func (h *handler) handleDeleteDB() error {
        h.assertAdminOnly()
        dbName := h.PathVar("olddb")

-       var bucket string
-
-       dbContext, _ := h.server.GetDatabase(h.ctx(), dbName)
-
-       if dbContext != nil {
-               bucket = dbContext.Bucket.GetName()
-       } else if h.server.persistentConfig {
-               bucket, _ = h.server.bucketNameFromDbName(dbName)
-       }
-
        if h.server.persistentConfig {
-               _, err := h.server.BootstrapContext.Connection.UpdateConfig(bucket, h.server.Config.Bootstrap.ConfigGroupID, func(rawBucketConfig []byte, rawBucketConfigCas uint64) (updatedConfig []byte, err error) {
-                       return nil, nil
-               })
-               if err != nil {
+               bucket, ok := h.server.bucketNameFromDbName(dbName)
+               if !ok {
+                       return base.HTTPErrorf(http.StatusNotFound, "no such database %q", dbName)
+               }
+
+               if _, err := h.server.BootstrapContext.Connection.UpdateConfig(bucket, h.server.Config.Bootstrap.ConfigGroupID,
+                       func(rawBucketConfig []byte, rawBucketConfigCas uint64) (updatedConfig []byte, err error) {
+                               return nil, nil
+                       },
+               ); err != nil {
                        return base.HTTPErrorf(http.StatusInternalServerError, "couldn't remove database %q from bucket %q: %s", base.MD(dbName), base.MD(bucket), err.Error())
                }
+
+               // we don't care whether the database is actually loaded or not when deleted - we report the delete either way...
+               _ = h.server.RemoveDatabase(h.ctx(), dbName)
+       } else {
+               if !h.server.RemoveDatabase(h.ctx(), dbName) {
+                       return base.HTTPErrorf(http.StatusNotFound, "qwerty no such database %q", dbName)
+               }
        }
-       if !h.server.RemoveDatabase(h.ctx(), dbName) && bucket == "" {
-               return base.HTTPErrorf(http.StatusNotFound, "no such database %q", dbName)
-       }
+
        _, _ = h.response.Write([]byte("{}"))
        return nil
 }
diff --git a/rest/config.go b/rest/config.go
index 7324f8a5..7f1dd52e 100644
--- a/rest/config.go
+++ b/rest/config.go
@@ -1428,6 +1428,14 @@ func (sc *ServerContext) fetchDatabase(ctx context.Context, dbName string) (foun
 }

 func (sc *ServerContext) bucketNameFromDbName(dbName string) (bucketName string, found bool) {
+       // see if we find the database in-memory first, otherwise fall back to scanning buckets for db configs
+       sc.lock.RLock()
+       dbc, ok := sc.databases_[dbName]
+       sc.lock.RUnlock()
+       if ok {
+               return dbc.Bucket.GetName(), true
+       }
+
        // Minimal representation of config struct to be tolerant of invalid database configurations where we still need to find a database name
        var cfgDbName struct {
                Name string `json:"name"`
```

Reply:
addressed in [f05a008](https://github.com/couchbase/sync_gateway/pull/5894/commits/f05a00887713e080a574e9f35fd52cf4b2e19014)